### PR TITLE
feat(admin): events for admin actions

### LINKED ISSUE #5444: onImpersonationStart / onImpersonationStop configuration hooks...
### Is this suited for github?

- [x] Yes, this is suited for github

### Is your feature request related to a problem? Please describe.

I want to log when impersonation sessions are started and ended in the database, but I'm not finding any API hooks or callbacks. 

### Describe the solution you'd like

I'd like to implement onImpersonationStarting() and onImpersonationStopping() type hooks in the admin configuration object on initialization. I'm sure there might be more sophisticated APIs that might allow blocking starting via an onBeforeImpersonationStarting(), but my logging use case would happily execute after starting and after ending so I can create a simple impersonation audit log.

### Describe alternatives you've considered

My work around is to instrument the code around the better-auth admin api calls.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ping-maxwell**
closes https://github.com/better-auth/better-auth/issues/5444
linear https://linear.app/better-auth/issue/ENG-546/feat-impersonation-hooks-for-admin-plugin

Normally hooks can achieve the most of it, but there are extra work required such as making sure the endpoint was a successful request and whatnot. Extra work for our end users. Additionally we believe this is a fair ask, so I've started work on this.

image

- [x] tests
- [x] docs 

This PR adds event options to the admin plugin to capture common events: impersonate start, impersonate end, ban user, unban user, user create, remove user.

Useful for:
- cleanups
- audit logging
- notifications / email sending

image

---
## Summary by cubic
Adds event callbacks to the admin plugin so apps can react to impersonation, bans, and user management. Addresses Linear ENG-546 by providing post-success impersonation events and reducing hook boilerplate.

- **New Features**
  - New admin.events: impersonateStart, impersonateEnd, ban, unban, userCreate, userRemove.
  - Event payloads include adminSession and the target user; adminSession may be null for server-initiated calls, while ban/unban always include a session.
  - Ban/unban refetch the user to pass updated fields (banned, banReason, banExpires).
  - Added documentation with examples and comprehensive tests.
  - Event errors do not block admin operations.

<sup>Written for commit 143c0044bcac3958adb2af0482cb1e2f1506d1dc. Summary will update on new commits.</sup>

### CODE REVIEW

**Comment by ashwingur** [comment]
Will this also allow for some checks before the action runs. For example if I want to prevent an admin from banning or impersonating another admin?

**Comment by ping-maxwell** [comment]
> Will this also allow for some checks before the action runs. For example if I want to prevent an admin from banning or impersonating another admin?

These are events that run after the action has succeeded, you should use hooks for that

File: packages/better-auth/src/plugins/admin/admin.ts [hunk 1]
```diff
@@ -1161,6 +1179,7 @@ export const admin = <O extends AdminOptions>(options?: O | undefined) => {
 						},
 						true,
 					);
+					await opts.events?.impersonateStart(ctx.context.session, targetUser);
```
[the remark this answers is not in the document]
Comment (ping-maxwell) [reply packages/better-auth/src/plugins/admin/admin.ts]: Fixed: admin session is now captured before `setSessionCookie` and passed to `impersonateStart`.

File: docs/content/docs/plugins/admin.mdx [hunk 2]
```diff
@@ -820,3 +820,161 @@ admin({
...
+```ts
+userRemove: async (adminSession, removedUser) => {
+  // Archive user data before deletion
+  await archiveUserData(removedUser.id);
+}
+```
+
+<Callout>
+  **Note:** All event callbacks are `async` functions and should return a `Promise<void>`.
+  If an event callback throws an error, it will not prevent the admin operation from completing.
```
[the remark this answers is not in the document]
Comment (ping-maxwell) [reply docs/content/docs/plugins/admin.mdx]: Fixed: updated the callout to document that unhandled errors in event callbacks will cause the admin operation to fail.

File: packages/better-auth/src/plugins/admin/admin.ts [hunk 3]
```diff
@@ -1138,6 +1156,7 @@ export const admin = <O extends AdminOptions>(options?: O | undefined) => {
 							message: ADMIN_ERROR_CODES.FAILED_TO_CREATE_USER,
 						});
 					}
+					await opts.events?.impersonateStart(ctx.context.session, targetUser);
```
[the remark this answers is not in the document]
Comment (ping-maxwell) [reply packages/better-auth/src/plugins/admin/admin.ts]: Fixed: `impersonateStart` is now triggered after cookie/session setup completes successfully, with the admin session captured beforehand.

File: packages/better-auth/src/plugins/admin/admin.test.ts [hunk 4]
```diff
@@ -1564,3 +1564,1014 @@ describe("edge cases: userId validation", async () => {
...
+			const [adminSession, bannedUser] = callArgs || [];
+			expect(adminSession?.user).toBeDefined();
+			expect(adminSession?.user.role).toBe("admin");
+			expect(adminSession?.session).toBeDefined();
+			expect(bannedUser?.banned).toBe(true);
+		});
+	});
+
+	describe("unban event", () => {
+		it("should call unban event when admin unban's a user", async () => {
```
[the remark this answers is not in the document]
Comment (ping-maxwell) [reply packages/better-auth/src/plugins/admin/admin.test.ts]: Fixed: corrected test description to use "unbans" instead of "unban's".

File: packages/better-auth/src/plugins/admin/routes.ts [hunk 5]
```diff
@@ -1630,6 +1684,15 @@ export const removeUser = (opts: AdminOptions) =>
...
+
+			// Call userRemove event before deleting the user
+			if (opts.events?.userRemove) {
+				const adminSession = {
+					user: session.user as User & Record<string, unknown>,
+					session: session.session as Session & Record<string, unknown>,
+				};
+				await opts.events.userRemove(adminSession, user as UserWithRole);
+			}
 			await ctx.context.internalAdapter.deleteUser(ctx.body.userId);
```
[the remark this answers is not in the document]
Comment (ping-maxwell) [reply packages/better-auth/src/plugins/admin/routes.ts]: Fixed in de74a8bd: userRemove now fires after deleteUser completes, and all event callbacks are wrapped in try/catch via invokeAdminEvent so a throwing callback cannot leave the request in a partial state.

File: packages/better-auth/src/plugins/admin/types.ts [hunk 6]
```diff
@@ -76,6 +76,82 @@ export interface AdminOptions {
...
+		unban: (
+			adminSession: {
+				user: User & Record<string, unknown>;
+				session: Session & Record<string, unknown>;
+			},
+			unbannedUser: UserWithRole,
+		) => Promise<void>;
+		/**
+		 * A callback when `admin.createUser` is called.
+		 */
```
[the remark this answers is not in the document]
Comment (ping-maxwell) [reply packages/better-auth/src/plugins/admin/types.ts]: Fixed in de74a8bd: removed the incorrect `| null` from impersonateEnd and userRemove types and updated docs to state adminSession is always defined for these callbacks.

File: packages/better-auth/src/plugins/admin/routes.ts [hunk 7]
```diff
@@ -1153,6 +1176,17 @@ export const banUser = (opts: AdminOptions) =>
...
+			// Call ban event
+			if (opts.events?.ban) {
+				await opts.events.ban(
+					{
+						user: session.user as User & Record<string, unknown>,
+						session: session.session as Session & Record<string, unknown>,
+					},
+					user as UserWithRole,
+				);
+			}
```
[the remark this answers is not in the document]
Comment (ping-maxwell) [reply packages/better-auth/src/plugins/admin/routes.ts:1213]: Fixed in de74a8bd: added invokeAdminEvent helper that catches and logs thrown errors without failing the admin operation, matching the PR contract. Docs callout updated accordingly.

File: docs/content/docs/plugins/admin.mdx [hunk 8]
```diff
@@ -845,3 +845,173 @@ admin({
...
+#### Event Callbacks
+
+##### `impersonateStart`
+
+Triggered when `admin.impersonateUser` is called.
+
+**Parameters:**
+
+* `adminSession`: The admin's session object (contains `user` and `session`). Will be `null` if the request is made from the server (without a session).
+* `impersonatedUser`: The user object that is being impersonated.
```
[the remark this answers is not in the document]
Comment (ping-maxwell) [reply docs/content/docs/plugins/admin.mdx:907]: Fixed in 4844ce323: removed the incorrect `| null` from the `impersonateStart` type and updated docs/examples to state that adminSession is always defined (impersonateUser requires an active admin session via adminMiddleware).
